customerio/customerio-ios
# refactor!: singleton API only way to use SDK now

# QA 

This PR touches almost the entire codebase and during my own QA testing, I have found a few bugs. So, a full QA walk through with the RH app is recommended. 

Things to test in particular:
* Install build of RH app *before* [`pr.78` build](https://github.com/customerio/RemoteHabits-iOS/pull/78). Run through the app a little bit to make sure some tasks are added to the background queue and profile and device token have been saved to app's data storage. Then, install `pr.78` build of RH. Open app and make sure that the previously identified profile, device token, and background queue tasks all run as expected. This tests that customers upgrading to this version of SDK will not lose data going from SDK v2 to v3. 
* Open app, do some tasks, force quit app, open it up again. The SDK initialization code has been changed so we want to make sure that it's working correctly when the RH opens and it initializes the SDK. 
* Push notifications and in-app messages are received. 
* Profile and device attributes get added successfully to Fly. RH app adds these automatically. We need to confirm that Fly receives them. 

---

Closes: https://github.com/customerio/issues/issues/6933

This PR removes the ability to use the iOS SDK in a non-singleton way. You used to be able to create 1+ instances of the SDK: `CustomerIO()` but now you have to use `CustomerIO.shared`. 

Notes: 
* This PR will trigger a breaking change to the SDK and will trigger a major version bump to v3.0
* This is a bigger PR since it's a larger refactor. I made comments on parts of the code I find interesting in the code review.
